docs/DesignDocs/AttestationAPIsWithFlexibleClaims.md

```diff
@@ -0,0 +1,478 @@
+   * @param[out] output_certificate_size The size of the buffer above.
+   *
+   * @return OE_OK on success
+   */
+  oe_result_t oe_get_attestation_certificate_with_evidence_v3(
+      const oe_uuid_t* format_id,
+      const unsigned char* subject_name,
+      const uint8_t* private_key,
+      size_t private_key_size,
+      const uint8_t* public_key,
+      size_t public_key_size,
+      const uint8_t* runtime_custom_claims_buffer,
+      size_t runtime_custom_claims_buffer_size,
+      const uint8_t* inittime_custom_claims_buffer,
+      size_t inittime_custom_claims_buffer_size,
+      const void* optional_parameters,
+      size_t optional_parameters_size,
+      uint8_t** output_certificate,
+      size_t* output_certificate_size);
+
+  ```
+  The API introduces four arguments that allow developers to optionally
+  pass in run- and init-time claims.
+  - `runtime_custom_claims_buffer`
+  - `runtime_custom_claims_buffer_size`
+  - `inittime_custom_claims_buffer`
+  - `inittime_custom_claims_buffer_size`
+
+- oe_verify_attestation_certificate_with_evidence_v3
+  ```c
+  /**
+   * oe_verify_attestation_certificate_with_evidence_v3
+   *
+   * This function is similar to
+   * oe_verify_attestation_certificate_with_evidence_v2 that has the same
+   * prototype but with the ability to verify a ceritifacte generated by
+   * oe_get_attestation_certificate_with_evidence_v3.
+   *
+   * @experimental
+   *
+   * @param[in] certificate_in_der A pointer to buffer holding certificate
+   * contents in DER format.
+   * @param[in] certificate_in_der_size The size of the buffer above.
+   * @param[in] endorsements_buffer A pointer to buffer holding endorsements.
+   * @param[in] endorsements_buffer_size The size of the buffer above.
+   * @param[in] policies A pointer to buffer holding policies.
+   * @param[in] policies_size The size of the buffer above.
+   * @param[out] claims A double-pointer to buffer holding claims.
+   * @param[out] claims_length The size of the buffer above.
+   * @retval OE_OK Successful validation.
+   * @retval OE_VERIFY_FAILED Quote failure.
+   * @retval OE_INVALID_PARAMETER One or more invalid parameters.
+   * @retval OE_FAILURE General failure.
+   * @retval Other appropriate error code.
+   */
+  oe_result_t oe_verify_attestation_certificate_with_evidence_v3(
+      const uint8_t* certificate_in_der,
+      size_t certificate_in_der_size,
+      const uint8_t* endorsements_buffer,
+      size_t endorsements_buffer_size,
+      const oe_policy_t* policies,
+      size_t policies_size,
+      oe_claim_t** claims,
+      size_t* claims_length);
+  ```
+  The API does not introduce changes to the function prototype. The purpose of
+  the API is to pair with `oe_get_attestation_certificate_with_evidence_v3`;
+  i.e., only the API is able to verify the certificated generated via
+  `oe_get_attestation_certificate_with_evidence_v3` with either `runtime_custom_claims_buffer`
+  or `inittime_custom_claims_buffer` supplied.
```

Comment:
Why can't we have both init claims and custom claims? And how do we verify runtime claims?

Reply:
Removed the word `either` that causes confusion.

The current design is that it's up to the verifier to implement the verification logic against the runtime claims output by the the `oe_verify_attestation_certificate_with_evidence_v3`.